### ISSUE TITLE
Example does not work on firefox

### ISSUE THREAD

**Issue by Edwardzyc** [jedisct1/libsodium.js#103]
Using firefox 55.0.2

TypeError: U.asm._sodium_init is undefined

Am I overlooking something? Works fine in chrome/edge

**Comment by jedisct1** [comment]
What demo are you talking about?

**Comment by Edwardzyc** [comment]
The example in the README

<script>
window.sodium = { onload: function(sodium) {
  alert(sodium.to_hex(sodium.crypto_generichash(64, 'test')));
}};
</script>
...
<script src="sodium.js" async></script>

**Comment by jedisct1** [comment]
Maybe because Firefox supports WASM and you need to preload the `.wasm` file as in https://github.com/jedisct1/libsodium.js/blob/master/wrapper/modinit.js ?

**Comment by jedisct1** [comment]
Do the tests at https://download.libsodium.org/libsodium.js/browsers-wasm-test/ work?

**Comment by jedisct1** [comment]
If you know how to fix this, that would be awesome.

Otherwise, I'm going to remove support for wasm. I don't know how to make it work reliably.

**Comment by Edwardzyc** [comment]
Haven't checked the tests.

I'll investigate for a fix. I was hoping it was just something on my end that I was doing wrong.

**Comment by Edwardzyc** [comment]
Can you provide more details on preloading the .wasm file or perhaps point to reading for it?

**Comment by jedisct1** [comment]
On preloading: if you don't, you get this assertions: https://github.com/kripken/embenchen/blob/7db82018dc79ffb0c371d85f8f934119f3f054c3/asm_v_wasm/wasm_zlib.c.js#L1607

```
on the web, we need the wasm binary to be preloaded and set on Module['wasmBinary']. emcc.py will do that for you when generating HTML (but not JS)
```

"(but not JS)" is the important and annoying part. Emscripten can generate a nice demo webpage that includes everything you need to automatically load the asmjs or wasm code. But outside of that demo page, you're on your own :/

The documentation page on Emscripten/WASM is also pretty scarse: https://github.com/kripken/emscripten/wiki/WebAssembly

**Comment by jedisct1** [comment]
I think the issue is that `require('libsodium-wasm.js')` doesn't return a ready-to-use module. It asynchronously loads the `.wasm` file, and only after that file is loaded, libsodium symbols are available.

An `onRuntimeInitialized` property can be attached to the module to invoke a callback after initialization.

This is great, but it means that we can't load `libsodium-wrappers` (or `libsodium-wasm`) as a module. This breaks the traditional way we use commonjs, amd, es6 modules, everything.

One possible solution is to return a promise instead. That doesn't seem too bad, and will eventually allow the compilation to happen in a worker as well. Promises are not supported by Internet Explorer, though (but it is by all versions of Edge as far as I know).

Shall we go that route, and force people to use a Promise polyfill for old browers?

**Comment by jedisct1** [comment]
Another solution could be to encode the wasm file using Base64. After compression, there is a ~24 Kb overhead in doing so. That's quite a lot, but it might remain an okay workaround.

**Comment by jedisct1** [comment]
I tried the base64 thing; it doesn't solve the issue. There's still an internal promise that has to be resolved, so `require('libsodium-wasm')._sodium_init()` is not going to work.

**Comment by jedisct1** [comment]
Superseded by #105
